src/dodal/common/data.py

```diff
@@ -0,0 +1,145 @@
+    Should be always incrementing, unique per-visit, co-ordinated with any other
+    scan engines.
+    """
+
+    collectionNumber: int
+
+
+class VisitServiceClientBase(ABC):
+    # VisitService API
+    """
+    Object responsible for I/O in determining collection number
+    """
+
+    @abstractmethod
+    async def create_new_collection(self) -> DataCollectionIdentifier:
+        """Create new collection"""
+
+    @abstractmethod
+    async def get_current_collection(self) -> DataCollectionIdentifier:
+        """Get current collection"""
+
+
+class VisitServiceClient(VisitServiceClientBase):
+    # VisitService API
+    """Client for the VisitService REST API exposed by the GDA Server to co-ordinate unique filenames.
+    While VisitService is embedded in GDA, url is likely to be `ixx-control:8088/api`
+    """
+
+    _url: str
+
+    def __init__(self, url: str) -> None:
+        self._url = url
+
+    async def create_new_collection(self) -> DataCollectionIdentifier:
+        async with ClientSession() as session:
+            async with session.post(f"{self._url}/numtracker") as response:
+                if response.status == 200:
+                    json = await response.json()
+                    return DataCollectionIdentifier.parse_obj(json)
+                else:
+                    raise Exception(response.status)
+
+    async def get_current_collection(self) -> DataCollectionIdentifier:
+        async with ClientSession() as session:
+            async with session.get(f"{self._url}/numtracker") as response:
+                if response.status == 200:
+                    json = await response.json()
+                    return DataCollectionIdentifier.parse_obj(json)
+                else:
+                    raise Exception(response.status)
+
+
+class LocalVisitServiceClient(VisitServiceClientBase):
+    # VisitService API
+    """Local or dummy impl of VisitService client to co-ordinate unique filenames."""
+
+    _count: int
+
+    def __init__(self) -> None:
+        self._count = 0
+
+    async def create_new_collection(self) -> DataCollectionIdentifier:
+        self._count += 1
+        return DataCollectionIdentifier(collectionNumber=self._count)
+
+    async def get_current_collection(self) -> DataCollectionIdentifier:
+        return DataCollectionIdentifier(collectionNumber=self._count)
+
+
+class StaticVisitDirectoryProvider(DirectoryProvider):
+    """
+    Gets information from a remote service to determine how their files should be named.
+    Supports a single Visit which should be passed as a Path relative to the root of the Detector IOC mounting.
+    i.e. for a Kubernetes IOC that mounts `/dls/ixx/data` as `/data`: root="/data/YYYY/visit/"
+    but for a Detector running with a view to the whole DLS filesystem, root="/dls/ixx/data/YYYY/visit"
+    Requires that all detectors are running with a mutual view on the filesystem.
```

Comment:
Is this actually true?

Reply:
The comment seems to suggest that there are both "detectors mounting the whole DLS filesystem" and "detectors mounting only the data directory". So my question was, is it the case that these detectors cannot and will not ever coexist on a beamline?

Ideally I'd like to find a nice way to _assert_ that's true at runtime, rather than waiting for an obscure error, but happy to spin that out into a separate issue...

Reply:
I think it needs to be a separate issue. I'm not sure how we can check other than trying to set a filepath to them and it failing.

Reply:
Created new issue to track this https://github.com/DiamondLightSource/dodal/issues/455